{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from Bio import SeqIO\n",
    "import getopt\n",
    "import os\n",
    "from BCBio import GFF\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "import random\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "help = '''\n",
    "    {script_name} -c com_port [-o output_file] [--loglevel level]\n",
    "\n",
    "    Reads the temperature data from a radio.  The temperature data is output in csv form.\n",
    "\n",
    "    examples:\n",
    "        Read table from radio attached to com4 and write the table to the file\n",
    "        output.csv.\n",
    "\n",
    "            {script_name} -c com4 -o output.csv\n",
    "\n",
    "        Read table from radio attached to com3 and write the table to stdout. \n",
    "        You can use IO redirection to send the contents where every you want.\n",
    "\n",
    "            # just print to the terminal \n",
    "            {script_name} -c com3\n",
    "\n",
    "            # redirect to another file\n",
    "            {script_name} -c com3 > somefile.csv\n",
    "\n",
    "            # filter out temperatures that are -100\n",
    "            {script_name} -c com3 | grep -v '^-100' \n",
    "\n",
    "\n",
    "    -c com_port\n",
    "    --com_port comport\n",
    "        Name of the COM port attached to the radio\n",
    "\n",
    "    -o output_file\n",
    "    --output output_file\n",
    "        If specified write the table data to the given file.  If not specified\n",
    "        the data will be written to stdout.\n",
    "\n",
    "    --loglevel critical | error | warning | info | debug | notset\n",
    "        Control the verbosity of the script by setting the log level.  Critical\n",
    "        is the least verbose and notset is the most verbose.\n",
    "\n",
    "        The default loglevel is {default_loglevel}.\n",
    "\n",
    "        These values correspond directly to the python logging module levels. \n",
    "        (i.e. https://docs.python.org/3/howto/logging.html#logging-levels)\n",
    "\n",
    "\n",
    "    -h \n",
    "    --help \n",
    "        print this message\n",
    "\n",
    "'''\n",
    "\n",
    "def usage():\n",
    "    print(help)\n",
    "    \n",
    "def rungetopts():\n",
    "    try:\n",
    "        opts, args = getopt.getopt(sys.argv[1:], \"a:qh\", [\"accession\", \"quiet\", \"help\"])\n",
    "    except getopt.GetoptError as err:\n",
    "        # print help information and exit:\n",
    "        print(err) # will print something like \"option -a not recognized\"\n",
    "        usage()\n",
    "        sys.exit(2)\n",
    "    accession = \"\"\n",
    "    for o, a in opts:\n",
    "            if o in (\"-h\", \"--help\"):\n",
    "                usage()\n",
    "                sys.exit()\n",
    "            elif o in (\"-a\", \"--accession\"):\n",
    "                accession = a\n",
    "            else:\n",
    "                assert False, \"unhandled option\"\n",
    "    if accession == \"\":\n",
    "        print(\"-a <accession> missing. For more help use -h\")\n",
    "        sys.exit(2)\n",
    "    return(accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getreaddepths(accession):\n",
    "    try:\n",
    "        read_depths_list = {}\n",
    "        for filename in os.listdir(\"/Users/thomasnicholson/phd/RNASeq/plot_files/cds_included/%s/\" % accession):\n",
    "            filesize = os.path.getsize(\"/Users/thomasnicholson/phd/RNASeq/plot_files/cds_included/%s/%s\" % (accession, filename))\n",
    "            if filesize == 0:\n",
    "                print(\"No data in %s\" % filename)\n",
    "                continue\n",
    "            plotFile  = open(os.path.join(\"/Users/thomasnicholson/phd/RNASeq/plot_files/cds_included/%s/\" % accession, filename), 'r')\n",
    "            print(filename)\n",
    "\n",
    "            read_depths = []\n",
    "            for line in plotFile:\n",
    "                words = line.rstrip()\n",
    "                words = words.split()\n",
    "                pos = words[0]\n",
    "                neg = words[1]\n",
    "                if pos > neg:\n",
    "                    read_depths.append(int(pos))\n",
    "                else:\n",
    "                    read_depths.append(int(neg))\n",
    "            read_depths_list[filename] = read_depths\n",
    "        print(\"making dataframe\")\n",
    "        df = pd.DataFrame(data = read_depths_list)\n",
    "\n",
    "\n",
    "        df['mean'] = df.iloc[:].mean(axis=1)\n",
    "        df['sum'] = df.sum(axis=1)\n",
    "        df['median'] = df.median(axis=1)\n",
    "        df['max'] = df.max(axis=1)\n",
    "        return df\n",
    "    except IOError:\n",
    "        print(\"Cannot open a file in /Users/thomasnicholson/phd/RNASeq/plot_files/cds_included/%s/\" % accession)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sRNA_read_depths(inFile, read_depths_df,accession):\n",
    "    outFile  = open(\"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/read_depths/%s_read_depths.txt\" % accession, 'w')\n",
    "    outFile.write(\"ID\\tstart\\tend\\tgroup\\tfeature\\tmean_mean\\tmean_median\\tmean_max\\tmedian_mean\\tmedian_median\\tmedian_max\\tmax_mean\\tmax_median\\tmax_max\\n\")\n",
    "    outFile.close()\n",
    "    outFile  = open(\"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/read_depths/%s_read_depths.txt\" % accession, 'a')\n",
    "\n",
    "    for line in inFile:\n",
    "        words = line.rstrip()\n",
    "        words = words.split(\"\\t\")\n",
    "        srna = words[-1]\n",
    "        start = words[2]\n",
    "        try:\n",
    "            start = int(start)\n",
    "        except ValueError:\n",
    "            continue\n",
    "        end = words[3]\n",
    "        end = int(end)\n",
    "        new_feature = words[8]\n",
    "        feature = words[1]\n",
    "        if new_feature == \"FALSE\":\n",
    "            srna_type = \"known\"\n",
    "        else:\n",
    "            srna_type = \"novel\"\n",
    "\n",
    "        subsetDF = read_depths_df[start:end]\n",
    "\n",
    "        mean_mean = subsetDF['mean'].mean()\n",
    "        mean_median = subsetDF['mean'].median()\n",
    "        mean_max = subsetDF['mean'].max()\n",
    "        median_mean = subsetDF['median'].mean()\n",
    "        median_median = subsetDF['median'].median()\n",
    "        median_max = subsetDF['median'].mean()\n",
    "        max_mean = subsetDF['max'].mean()\n",
    "        max_median = subsetDF['max'].median()\n",
    "        max_max = subsetDF['max'].max()\n",
    "\n",
    "        outFile.write(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (srna, start, end, srna_type, feature, mean_mean, mean_median, mean_max, median_mean, median_median, median_max, max_mean, max_median, max_max))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# accession = rungetopts()\n",
    "accession = \"GCA_000006765.1\"\n",
    "random = True\n",
    "print(random)\n",
    "print(\"Reading files\")\n",
    "try:\n",
    "    if random == False:\n",
    "        inFile = open(\"/Users/thomasnicholson/phd/RNASeq/new_calls/%s_new_calls.txt\" % accession, 'r')\n",
    "    else:\n",
    "        #inFile = open(\"/Users/thomasnicholson/phd/RNASeq/new_calls/random/python_version_1/%s_random_new_calls.txt\" % accession, 'r')\n",
    "        print(\"Running read_csv\")\n",
    "        inDF = pd.read_csv(\"/Users/thomasnicholson/phd/RNASeq/new_calls/random/python_version_1/%s_random_new_calls.txt\" % accession, sep=\"\\t\")\n",
    "except IOError:\n",
    "    print(\"/Users/thomasnicholson/phd/RNASeq/new_calls/%s_new_calls.txt not found\" % accession)\n",
    "    sys.exit(2)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(inDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_depths_list = {}\n",
    "df = None\n",
    "for filename in os.listdir(\"/Users/thomasnicholson/phd/RNASeq/plot_files/cds_included/%s/\" % accession):\n",
    "        filesize = os.path.getsize(\"/Users/thomasnicholson/phd/RNASeq/plot_files/cds_included/%s/%s\" % (accession, filename))\n",
    "        if filesize == 0:\n",
    "            print(\"No data in %s\" % filename)\n",
    "            continue\n",
    "        plotFile  = pd.read_csv(os.path.join(\"/Users/thomasnicholson/phd/RNASeq/plot_files/cds_included/%s/\" % accession, filename), sep = '\\t', header=None)\n",
    "        print(filename)\n",
    "        plotFile['selected'] = plotFile.iloc[:].max(axis=1)\n",
    "        tmpDf = plotFile.iloc[:,2]\n",
    "        if df is not None:\n",
    "            df = pd.concat([df.reset_index(drop=True), tmpDf], axis=1)\n",
    "        else:\n",
    "            df = tmpDf\n",
    "#         read_depths = []\n",
    "#         for line in plotFile:\n",
    "#             words = line.rstrip()\n",
    "#             words = words.split()\n",
    "#             pos = words[0]\n",
    "#             neg = words[1]\n",
    "#             if pos > neg:\n",
    "#                 read_depths.append(int(pos))\n",
    "#             else:\n",
    "#                 read_depths.append(int(neg))\n",
    "#         read_depths_list[filename] = read_depths\n",
    "# print(\"making dataframe\")\n",
    "# df = pd.DataFrame(data = read_depths_list)\n",
    "\n",
    "dfOut = df\n",
    "dfOut['mean'] = df.iloc[:].mean(axis=1)\n",
    "dfOut['median'] = df.median(axis=1)\n",
    "dfOut['max'] = df.max(axis=1)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# display(read_depths_list)\n",
    "display(dfOut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"getting read depths\")\n",
    "read_depths_df = getreaddepths(accession)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(read_depths_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"writing file\")\n",
    "\n",
    "sRNA_read_depths(inFile, dfOut, accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def openNHMMER(nhmmername):\n",
    "    nhmmerDF = pd.read_csv(nhmmername, delim_whitespace=True, header=None, comment='#')\n",
    "    nhmmerDF.columns = [\"target_name\", \"accession\", \"query_name\", \"accession_2\", \"hmmfrom\", \"hmmto\", \"alifrom\", \"alito\", \"envfrom\", \"envto\", \"sq_len\", \"strand\", \"E_value\", \"score\", \"bias\", \"description_of_target\"]\n",
    "    nhmmerDF[[\"ID\", \"descriptors\"]] = nhmmerDF.target_name.str.split(\"[\", expand = True)\n",
    "    nhmmerDF[[\"ID_2\", \"descriptors_2\"]] = nhmmerDF.query_name.str.split(\"[\", expand = True)\n",
    "    d = nhmmerDF.groupby('ID')['ID_2'].apply(list).to_dict()\n",
    "    return(d)\n",
    "\n",
    "def openReadDepths(readdepthsname, d):\n",
    "    readdepthsDF = pd.read_csv(readdepthsname, sep = \"\\t\", comment='#')\n",
    "    readdepthsDF = readdepthsDF[readdepthsDF['ID'] != \"ID\"]\n",
    "    readdepthsDF[[\"mean_value\", \"mean_decimal\"]] = readdepthsDF.max_mean.str.split(\".\", expand = True)\n",
    "    readdepthsDF[[\"median_value\", \"median_decimal\"]] = readdepthsDF.max_median.str.split(\".\", expand = True)\n",
    "    readdepthsDF[[\"max_value\", \"max_decimal\"]] = readdepthsDF.max_max.str.split(\".\", expand = True)\n",
    "    readdepthsDF[['mean_value', 'median_value', 'max_value']] = readdepthsDF.loc[:,['mean_value', 'median_value', 'max_value']].apply(pd.to_numeric)\n",
    "    idList = list(d.keys())\n",
    "    readdepthsKept = readdepthsDF[readdepthsDF['ID'].isin(idList)]\n",
    "    return(readdepthsKept)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def writeReadDepths(outname, readDepths, d):\n",
    "    seen = []\n",
    "    d2 = {}\n",
    "    i = 0\n",
    "\n",
    "    outFile = open(outname, \"w\")\n",
    "    outFile.write(\"ID\\tmean_mean\\tmean_median\\tmean_max\\tmedian_mean\\tmedian_median\\tmedian_max\\tmax_mean\\tmax_median\\tmax_max\\tID_2\\n\")\n",
    "    outFile.close()\n",
    "    outFile = open(outname, \"a\")\n",
    "    values = []\n",
    "    for key in d:\n",
    "    #     print(i)\n",
    "    #     i += 1\n",
    "    #     if i > 100:\n",
    "    #         break\n",
    "    #     if key in seen:\n",
    "    #         continue\n",
    "    #     print(key)\n",
    "    #     print(seen)\n",
    "        values = d[key]\n",
    "        seen.append(values)\n",
    "        df = readDepths[readDepths['ID'].isin(values)]\n",
    "        #     print(df['mean_value'].dtypes)\n",
    "        \n",
    "    #     print(df['mean_value'].dtypes)\n",
    "        mean_mean = df['mean_value'].mean()\n",
    "        mean_median = df['mean_value'].median()\n",
    "        mean_max = df['mean_value'].max()\n",
    "        median_mean = df['median_value'].mean()\n",
    "        median_median = df['median_value'].median()\n",
    "        median_max = df['median_value'].max()\n",
    "        max_mean = df['max_value'].mean()\n",
    "        max_median = df['max_value'].median()\n",
    "        max_max = df['max_value'].max()\n",
    "    #     print(key)\n",
    "    #     print(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (key,mean_mean,mean_median,mean_max,median_mean,median_median,median_max,max_mean,max_median,max_max,values))\n",
    "        outFile.write(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (key,mean_mean,mean_median,mean_max,median_mean,median_median,median_max,max_mean,max_median,max_max,values))\n",
    "    outFile.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = openNHMMER(nhmmername = \"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/positive_control.tbl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "readDepthsKept = openReadDepths(readdepthsname  = \"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/read_depths.txt\", d = d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "writeReadDepths(outname = \"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/positive_control_read_depths.txt\", readDepths = readDepthsKept, d = d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def openNHMMER(nhmmername):\n",
    "    nhmmerDF = pd.read_csv(nhmmername, delim_whitespace=True, header=None, comment='#')\n",
    "    nhmmerDF.columns = [\"target_name\", \"accession\", \"query_name\", \"accession_2\", \"hmmfrom\", \"hmmto\", \"alifrom\", \"alito\", \"envfrom\", \"envto\", \"sq_len\", \"strand\", \"E_value\", \"score\", \"bias\", \"description_of_target\"]\n",
    "    nhmmerDF[[\"ID\", \"descriptors\"]] = nhmmerDF.target_name.str.split(\"[\", expand = True)\n",
    "    nhmmerDF[[\"ID_2\", \"descriptors_2\"]] = nhmmerDF.query_name.str.split(\"[\", expand = True)\n",
    "    d = nhmmerDF.groupby('ID')['ID_2'].apply(list).to_dict()\n",
    "    return(d, nhmmerDF)\n",
    "def getreaddepths(accession):\n",
    "    try:\n",
    "        df = None\n",
    "        for filename in os.listdir(\"/Users/thomasnicholson/phd/RNASeq/plot_files/%s/\" % accession):\n",
    "            filesize = os.path.getsize(\n",
    "                \"/Users/thomasnicholson/phd/RNASeq/plot_files/%s/%s\" % (accession, filename))\n",
    "            if filesize == 0:\n",
    "                print(\"No data in %s\" % filename)\n",
    "                continue\n",
    "            plotFile = pd.read_csv(\n",
    "                os.path.join(\"/Users/thomasnicholson/phd/RNASeq/plot_files/%s/\" % accession, filename),\n",
    "                sep='\\t', header=None)\n",
    "            print(filename)\n",
    "            plotFile['selected'] = plotFile.iloc[:].max(axis=1)\n",
    "            tmpDf = plotFile.iloc[:, 2]\n",
    "            if df is not None:\n",
    "                df = pd.concat([df.reset_index(drop=True), tmpDf], axis=1)\n",
    "            else:\n",
    "                df = tmpDf\n",
    "        dfOut = df\n",
    "        dfOut['mean'] = df.iloc[:].mean(axis=1)\n",
    "        dfOut['median'] = df.median(axis=1)\n",
    "        dfOut['max'] = df.max(axis=1)\n",
    "        return dfOut\n",
    "\n",
    "    except IOError:\n",
    "        print(\"Cannot open a file in /Users/thomasnicholson/phd/RNASeq/plot_files/%s/\" % accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "d, nhmmerDF = openNHMMER(nhmmername = \"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/predicted.tbl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR3990750_ncRNA.plot\n",
      "SRR3990749_ncRNA.plot\n",
      "SRR3990735_ncRNA.plot\n",
      "SRR3990733_ncRNA.plot\n",
      "SRR3990734_ncRNA.plot\n",
      "SRR3990742_ncRNA.plot\n"
     ]
    }
   ],
   "source": [
    "plotFiles = getreaddepths(\"GCA_002355295.1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>query_name</th>\n",
       "      <th>accession_2</th>\n",
       "      <th>hmmfrom</th>\n",
       "      <th>hmmto</th>\n",
       "      <th>alifrom</th>\n",
       "      <th>alito</th>\n",
       "      <th>envfrom</th>\n",
       "      <th>envto</th>\n",
       "      <th>sq_len</th>\n",
       "      <th>strand</th>\n",
       "      <th>E_value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>description_of_target</th>\n",
       "      <th>ID</th>\n",
       "      <th>descriptors</th>\n",
       "      <th>ID_2</th>\n",
       "      <th>descriptors_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GCA_000006765.1_1[101-227,+,novel,ncRNA,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_1[101-227,+,novel,ncRNA,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>126</td>\n",
       "      <td>1</td>\n",
       "      <td>126</td>\n",
       "      <td>1</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>+</td>\n",
       "      <td>4.400000e-35</td>\n",
       "      <td>120.1</td>\n",
       "      <td>1.5</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_1</td>\n",
       "      <td>101-227,+,novel,ncRNA,1]</td>\n",
       "      <td>GCA_000006765.1_1</td>\n",
       "      <td>101-227,+,novel,ncRNA,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GCA_000006765.1_2[8466-8521,+,novel,ncRNA,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_2[8466-8521,+,novel,ncRNA,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>54</td>\n",
       "      <td>1</td>\n",
       "      <td>54</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>+</td>\n",
       "      <td>1.000000e-11</td>\n",
       "      <td>44.9</td>\n",
       "      <td>2.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_2</td>\n",
       "      <td>8466-8521,+,novel,ncRNA,1]</td>\n",
       "      <td>GCA_000006765.1_2</td>\n",
       "      <td>8466-8521,+,novel,ncRNA,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GCA_000006765.1_7[53652-53741,+,novel,ncRNA,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_7[53652-53741,+,novel,ncRNA,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>89</td>\n",
       "      <td>1</td>\n",
       "      <td>89</td>\n",
       "      <td>1</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>+</td>\n",
       "      <td>8.400000e-23</td>\n",
       "      <td>80.3</td>\n",
       "      <td>1.8</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_7</td>\n",
       "      <td>53652-53741,+,novel,ncRNA,1]</td>\n",
       "      <td>GCA_000006765.1_7</td>\n",
       "      <td>53652-53741,+,novel,ncRNA,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>GCA_000006765.1_8[55057-55480,+,novel,ncRNA,0....</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_8[55057-55480,+,novel,ncRNA,0....</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>422</td>\n",
       "      <td>1</td>\n",
       "      <td>422</td>\n",
       "      <td>1</td>\n",
       "      <td>423</td>\n",
       "      <td>423</td>\n",
       "      <td>+</td>\n",
       "      <td>4.500000e-130</td>\n",
       "      <td>432.7</td>\n",
       "      <td>10.3</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_8</td>\n",
       "      <td>55057-55480,+,novel,ncRNA,0.517730496453901]</td>\n",
       "      <td>GCA_000006765.1_8</td>\n",
       "      <td>55057-55480,+,novel,ncRNA,0.517730496453901]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>GCA_000006765.1_9[56034-56255,+,novel,ncRNA,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_9[56034-56255,+,novel,ncRNA,1]</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>221</td>\n",
       "      <td>1</td>\n",
       "      <td>221</td>\n",
       "      <td>1</td>\n",
       "      <td>221</td>\n",
       "      <td>221</td>\n",
       "      <td>+</td>\n",
       "      <td>1.800000e-66</td>\n",
       "      <td>223.1</td>\n",
       "      <td>1.4</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000006765.1_9</td>\n",
       "      <td>56034-56255,+,novel,ncRNA,1]</td>\n",
       "      <td>GCA_000006765.1_9</td>\n",
       "      <td>56034-56255,+,novel,ncRNA,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>71785</td>\n",
       "      <td>GCA_900243355.1_879[7753320-7753459,+,novel,nc...</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_900243355.1_879[7753320-7753459,+,novel,nc...</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>139</td>\n",
       "      <td>1</td>\n",
       "      <td>139</td>\n",
       "      <td>1</td>\n",
       "      <td>139</td>\n",
       "      <td>139</td>\n",
       "      <td>+</td>\n",
       "      <td>1.900000e-37</td>\n",
       "      <td>128.4</td>\n",
       "      <td>8.7</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_900243355.1_879</td>\n",
       "      <td>7753320-7753459,+,novel,ncRNA,0.76978417266187]</td>\n",
       "      <td>GCA_900243355.1_879</td>\n",
       "      <td>7753320-7753459,+,novel,ncRNA,0.76978417266187]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>71786</td>\n",
       "      <td>GCA_000017745.1_1905[4201112-4203303,+,novel,n...</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_900243355.1_879[7753320-7753459,+,novel,nc...</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "      <td>137</td>\n",
       "      <td>2178</td>\n",
       "      <td>2054</td>\n",
       "      <td>2191</td>\n",
       "      <td>2052</td>\n",
       "      <td>2191</td>\n",
       "      <td>-</td>\n",
       "      <td>1.300000e-11</td>\n",
       "      <td>46.1</td>\n",
       "      <td>2.2</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000017745.1_1905</td>\n",
       "      <td>4201112-4203303,+,novel,ncRNA,1]</td>\n",
       "      <td>GCA_900243355.1_879</td>\n",
       "      <td>7753320-7753459,+,novel,ncRNA,0.76978417266187]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>71787</td>\n",
       "      <td>GCA_900243355.1_880[7753767-7753860,+,novel,nc...</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_900243355.1_880[7753767-7753860,+,novel,nc...</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>92</td>\n",
       "      <td>1</td>\n",
       "      <td>92</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>+</td>\n",
       "      <td>1.200000e-23</td>\n",
       "      <td>83.3</td>\n",
       "      <td>4.1</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_900243355.1_880</td>\n",
       "      <td>7753767-7753860,+,novel,ncRNA,1]</td>\n",
       "      <td>GCA_900243355.1_880</td>\n",
       "      <td>7753767-7753860,+,novel,ncRNA,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>71788</td>\n",
       "      <td>GCA_000017745.1_1905[4201112-4203303,+,novel,n...</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_900243355.1_880[7753767-7753860,+,novel,nc...</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>91</td>\n",
       "      <td>1741</td>\n",
       "      <td>1653</td>\n",
       "      <td>1743</td>\n",
       "      <td>1651</td>\n",
       "      <td>2191</td>\n",
       "      <td>-</td>\n",
       "      <td>4.100000e-06</td>\n",
       "      <td>28.6</td>\n",
       "      <td>1.2</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_000017745.1_1905</td>\n",
       "      <td>4201112-4203303,+,novel,ncRNA,1]</td>\n",
       "      <td>GCA_900243355.1_880</td>\n",
       "      <td>7753767-7753860,+,novel,ncRNA,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>71789</td>\n",
       "      <td>GCA_900243355.1_881[7755222-7755353,+,novel,nc...</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_900243355.1_881[7755222-7755353,+,novel,nc...</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>131</td>\n",
       "      <td>1</td>\n",
       "      <td>131</td>\n",
       "      <td>1</td>\n",
       "      <td>131</td>\n",
       "      <td>131</td>\n",
       "      <td>+</td>\n",
       "      <td>8.400000e-36</td>\n",
       "      <td>122.7</td>\n",
       "      <td>4.7</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_900243355.1_881</td>\n",
       "      <td>7755222-7755353,+,novel,ncRNA,1]</td>\n",
       "      <td>GCA_900243355.1_881</td>\n",
       "      <td>7755222-7755353,+,novel,ncRNA,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71790 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             target_name accession  \\\n",
       "0             GCA_000006765.1_1[101-227,+,novel,ncRNA,1]         -   \n",
       "1           GCA_000006765.1_2[8466-8521,+,novel,ncRNA,1]         -   \n",
       "2         GCA_000006765.1_7[53652-53741,+,novel,ncRNA,1]         -   \n",
       "3      GCA_000006765.1_8[55057-55480,+,novel,ncRNA,0....         -   \n",
       "4         GCA_000006765.1_9[56034-56255,+,novel,ncRNA,1]         -   \n",
       "...                                                  ...       ...   \n",
       "71785  GCA_900243355.1_879[7753320-7753459,+,novel,nc...         -   \n",
       "71786  GCA_000017745.1_1905[4201112-4203303,+,novel,n...         -   \n",
       "71787  GCA_900243355.1_880[7753767-7753860,+,novel,nc...         -   \n",
       "71788  GCA_000017745.1_1905[4201112-4203303,+,novel,n...         -   \n",
       "71789  GCA_900243355.1_881[7755222-7755353,+,novel,nc...         -   \n",
       "\n",
       "                                              query_name accession_2  hmmfrom  \\\n",
       "0             GCA_000006765.1_1[101-227,+,novel,ncRNA,1]           -        1   \n",
       "1           GCA_000006765.1_2[8466-8521,+,novel,ncRNA,1]           -        1   \n",
       "2         GCA_000006765.1_7[53652-53741,+,novel,ncRNA,1]           -        1   \n",
       "3      GCA_000006765.1_8[55057-55480,+,novel,ncRNA,0....           -        1   \n",
       "4         GCA_000006765.1_9[56034-56255,+,novel,ncRNA,1]           -        1   \n",
       "...                                                  ...         ...      ...   \n",
       "71785  GCA_900243355.1_879[7753320-7753459,+,novel,nc...           -        1   \n",
       "71786  GCA_900243355.1_879[7753320-7753459,+,novel,nc...           -       13   \n",
       "71787  GCA_900243355.1_880[7753767-7753860,+,novel,nc...           -        1   \n",
       "71788  GCA_900243355.1_880[7753767-7753860,+,novel,nc...           -        3   \n",
       "71789  GCA_900243355.1_881[7755222-7755353,+,novel,nc...           -        1   \n",
       "\n",
       "       hmmto  alifrom  alito  envfrom  envto  sq_len strand        E_value  \\\n",
       "0        126        1    126        1    126     126      +   4.400000e-35   \n",
       "1         54        1     54        1     55      55      +   1.000000e-11   \n",
       "2         89        1     89        1     89      89      +   8.400000e-23   \n",
       "3        422        1    422        1    423     423      +  4.500000e-130   \n",
       "4        221        1    221        1    221     221      +   1.800000e-66   \n",
       "...      ...      ...    ...      ...    ...     ...    ...            ...   \n",
       "71785    139        1    139        1    139     139      +   1.900000e-37   \n",
       "71786    137     2178   2054     2191   2052    2191      -   1.300000e-11   \n",
       "71787     92        1     92        1     93      93      +   1.200000e-23   \n",
       "71788     91     1741   1653     1743   1651    2191      -   4.100000e-06   \n",
       "71789    131        1    131        1    131     131      +   8.400000e-36   \n",
       "\n",
       "       score  bias description_of_target                    ID  \\\n",
       "0      120.1   1.5                     -     GCA_000006765.1_1   \n",
       "1       44.9   2.1                     -     GCA_000006765.1_2   \n",
       "2       80.3   1.8                     -     GCA_000006765.1_7   \n",
       "3      432.7  10.3                     -     GCA_000006765.1_8   \n",
       "4      223.1   1.4                     -     GCA_000006765.1_9   \n",
       "...      ...   ...                   ...                   ...   \n",
       "71785  128.4   8.7                     -   GCA_900243355.1_879   \n",
       "71786   46.1   2.2                     -  GCA_000017745.1_1905   \n",
       "71787   83.3   4.1                     -   GCA_900243355.1_880   \n",
       "71788   28.6   1.2                     -  GCA_000017745.1_1905   \n",
       "71789  122.7   4.7                     -   GCA_900243355.1_881   \n",
       "\n",
       "                                           descriptors                 ID_2  \\\n",
       "0                             101-227,+,novel,ncRNA,1]    GCA_000006765.1_1   \n",
       "1                           8466-8521,+,novel,ncRNA,1]    GCA_000006765.1_2   \n",
       "2                         53652-53741,+,novel,ncRNA,1]    GCA_000006765.1_7   \n",
       "3         55057-55480,+,novel,ncRNA,0.517730496453901]    GCA_000006765.1_8   \n",
       "4                         56034-56255,+,novel,ncRNA,1]    GCA_000006765.1_9   \n",
       "...                                                ...                  ...   \n",
       "71785  7753320-7753459,+,novel,ncRNA,0.76978417266187]  GCA_900243355.1_879   \n",
       "71786                 4201112-4203303,+,novel,ncRNA,1]  GCA_900243355.1_879   \n",
       "71787                 7753767-7753860,+,novel,ncRNA,1]  GCA_900243355.1_880   \n",
       "71788                 4201112-4203303,+,novel,ncRNA,1]  GCA_900243355.1_880   \n",
       "71789                 7755222-7755353,+,novel,ncRNA,1]  GCA_900243355.1_881   \n",
       "\n",
       "                                         descriptors_2  \n",
       "0                             101-227,+,novel,ncRNA,1]  \n",
       "1                           8466-8521,+,novel,ncRNA,1]  \n",
       "2                         53652-53741,+,novel,ncRNA,1]  \n",
       "3         55057-55480,+,novel,ncRNA,0.517730496453901]  \n",
       "4                         56034-56255,+,novel,ncRNA,1]  \n",
       "...                                                ...  \n",
       "71785  7753320-7753459,+,novel,ncRNA,0.76978417266187]  \n",
       "71786  7753320-7753459,+,novel,ncRNA,0.76978417266187]  \n",
       "71787                 7753767-7753860,+,novel,ncRNA,1]  \n",
       "71788                 7753767-7753860,+,novel,ncRNA,1]  \n",
       "71789                 7755222-7755353,+,novel,ncRNA,1]  \n",
       "\n",
       "[71790 rows x 20 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(nhmmerDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "newCalls = pd.read_csv(\"/Users/thomasnicholson/phd/RNASeq/new_calls/GCA_002355295.1_new_calls.txt\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>file_names</th>\n",
       "      <th>row_numbers</th>\n",
       "      <th>prop_overlap</th>\n",
       "      <th>new_feature</th>\n",
       "      <th>number_of_rnaseq_files</th>\n",
       "      <th>score</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>1462</td>\n",
       "      <td>2053</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990733_sra_calls.gff,SRR3990734_sra_calls....</td>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>GCA_002355295.1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>5075</td>\n",
       "      <td>5242</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR3990735_sra_calls.gff,SRR3990750_sra_calls....</td>\n",
       "      <td>13,14,15,16,17</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>14319</td>\n",
       "      <td>14514</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990742_sra_calls.gff,SRR3990750_sra_calls....</td>\n",
       "      <td>18,19,20,21,22,23,24</td>\n",
       "      <td>0.856410</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>GCA_002355295.1_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>19048</td>\n",
       "      <td>19330</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990735_sra_calls.gff,SRR3990749_sra_calls....</td>\n",
       "      <td>25,26,27,28,29,30</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>22748</td>\n",
       "      <td>22887</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR3990750_sra_calls.gff,SRR3990735_sra_calls....</td>\n",
       "      <td>31,32,33,34,35,36,37,38</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>GCA_002355295.1_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>576</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>6183676</td>\n",
       "      <td>6183910</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990750_sra_calls.gff,SRR3990735_sra_calls....</td>\n",
       "      <td>3321,3322,3323,3324,3325</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>577</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>6188032</td>\n",
       "      <td>6188136</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR3990735_sra_calls.gff,SRR3990734_sra_calls.gff</td>\n",
       "      <td>3326,3327</td>\n",
       "      <td>0.913462</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>GCA_002355295.1_578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>578</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>6196182</td>\n",
       "      <td>6196368</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990735_sra_calls.gff,SRR3990750_sra_calls....</td>\n",
       "      <td>3328,3329,3330,3331,3332,3333,3334,3335,3336,3337</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>579</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>6244744</td>\n",
       "      <td>6244848</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR3990734_sra_calls.gff,SRR3990749_sra_calls....</td>\n",
       "      <td>3338,3339,3340,3341,3342,3343</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>GCA_002355295.1_580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>580</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>6255083</td>\n",
       "      <td>6255329</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR3990735_sra_calls.gff,SRR3990734_sra_calls....</td>\n",
       "      <td>3344,3345,3346</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>GCA_002355295.1_581</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>581 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sequence feature    start      end strand  \\\n",
       "0    NZ_CP023465.1   ncRNA     1462     2053      +   \n",
       "1    NZ_CP023465.1   ncRNA     5075     5242      -   \n",
       "2    NZ_CP023465.1   ncRNA    14319    14514      +   \n",
       "3    NZ_CP023465.1   ncRNA    19048    19330      +   \n",
       "4    NZ_CP023465.1   ncRNA    22748    22887      -   \n",
       "..             ...     ...      ...      ...    ...   \n",
       "576  NZ_CP023465.1   ncRNA  6183676  6183910      +   \n",
       "577  NZ_CP023465.1   ncRNA  6188032  6188136      -   \n",
       "578  NZ_CP023465.1   ncRNA  6196182  6196368      +   \n",
       "579  NZ_CP023465.1   ncRNA  6244744  6244848      -   \n",
       "580  NZ_CP023465.1   ncRNA  6255083  6255329      -   \n",
       "\n",
       "                                            file_names  \\\n",
       "0    SRR3990733_sra_calls.gff,SRR3990734_sra_calls....   \n",
       "1    SRR3990735_sra_calls.gff,SRR3990750_sra_calls....   \n",
       "2    SRR3990742_sra_calls.gff,SRR3990750_sra_calls....   \n",
       "3    SRR3990735_sra_calls.gff,SRR3990749_sra_calls....   \n",
       "4    SRR3990750_sra_calls.gff,SRR3990735_sra_calls....   \n",
       "..                                                 ...   \n",
       "576  SRR3990750_sra_calls.gff,SRR3990735_sra_calls....   \n",
       "577  SRR3990735_sra_calls.gff,SRR3990734_sra_calls.gff   \n",
       "578  SRR3990735_sra_calls.gff,SRR3990750_sra_calls....   \n",
       "579  SRR3990734_sra_calls.gff,SRR3990749_sra_calls....   \n",
       "580  SRR3990735_sra_calls.gff,SRR3990734_sra_calls....   \n",
       "\n",
       "                                           row_numbers  prop_overlap  \\\n",
       "0                           1,2,3,4,5,6,7,8,9,10,11,12      1.000000   \n",
       "1                                       13,14,15,16,17      1.000000   \n",
       "2                                 18,19,20,21,22,23,24      0.856410   \n",
       "3                                    25,26,27,28,29,30      1.000000   \n",
       "4                              31,32,33,34,35,36,37,38      1.000000   \n",
       "..                                                 ...           ...   \n",
       "576                           3321,3322,3323,3324,3325      1.000000   \n",
       "577                                          3326,3327      0.913462   \n",
       "578  3328,3329,3330,3331,3332,3333,3334,3335,3336,3337      1.000000   \n",
       "579                      3338,3339,3340,3341,3342,3343      1.000000   \n",
       "580                                     3344,3345,3346      1.000000   \n",
       "\n",
       "     new_feature  number_of_rnaseq_files score                   id  \n",
       "0           True                      12     7    GCA_002355295.1_1  \n",
       "1           True                       5     2    GCA_002355295.1_2  \n",
       "2           True                       7     3    GCA_002355295.1_3  \n",
       "3           True                       6     2    GCA_002355295.1_4  \n",
       "4           True                       8     4    GCA_002355295.1_5  \n",
       "..           ...                     ...   ...                  ...  \n",
       "576         True                       5     2  GCA_002355295.1_577  \n",
       "577         True                       2     3  GCA_002355295.1_578  \n",
       "578         True                      10     2  GCA_002355295.1_579  \n",
       "579         True                       6     6  GCA_002355295.1_580  \n",
       "580         True                       3     1  GCA_002355295.1_581  \n",
       "\n",
       "[581 rows x 12 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(newCalls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "idName = \"GCA_002355295.1_131\"\n",
    "values = d[idName]\n",
    "df = newCalls[newCalls['id'].isin(values)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>file_names</th>\n",
       "      <th>row_numbers</th>\n",
       "      <th>prop_overlap</th>\n",
       "      <th>new_feature</th>\n",
       "      <th>number_of_rnaseq_files</th>\n",
       "      <th>score</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>14319</td>\n",
       "      <td>14514</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990742_sra_calls.gff,SRR3990750_sra_calls....</td>\n",
       "      <td>18,19,20,21,22,23,24</td>\n",
       "      <td>0.856410</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>GCA_002355295.1_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>164658</td>\n",
       "      <td>164831</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990749_sra_calls.gff,SRR3990734_sra_calls....</td>\n",
       "      <td>128,129,130,131,132</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>249532</td>\n",
       "      <td>249705</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR3990742_sra_calls.gff,SRR3990749_sra_calls....</td>\n",
       "      <td>170,171,172,173,174,175</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>37</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>328878</td>\n",
       "      <td>329088</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990749_sra_calls.gff,SRR3990750_sra_calls....</td>\n",
       "      <td>245,246,247,248,249,250,251,252,253</td>\n",
       "      <td>0.709524</td>\n",
       "      <td>True</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>38</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>329138</td>\n",
       "      <td>329442</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990749_sra_calls.gff,SRR3990742_sra_calls....</td>\n",
       "      <td>254,255,256,257,258</td>\n",
       "      <td>0.361842</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>536</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>5655550</td>\n",
       "      <td>5655801</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990734_sra_calls.gff,SRR3990735_sra_calls....</td>\n",
       "      <td>3086,3087,3088,3089,3090,3091</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>GCA_002355295.1_537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>537</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>5655862</td>\n",
       "      <td>5656131</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR3990735_sra_calls.gff,SRR3990733_sra_calls....</td>\n",
       "      <td>3092,3093,3094,3095,3096,3097,3098,3099,3100,3...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>550</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>5807459</td>\n",
       "      <td>5807607</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR3990750_sra_calls.gff,SRR3990734_sra_calls....</td>\n",
       "      <td>3178,3179,3180,3181,3182,3183,3184,3185,3186</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>567</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>6074106</td>\n",
       "      <td>6074294</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR3990749_sra_calls.gff,SRR3990734_sra_calls.gff</td>\n",
       "      <td>3283,3284</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>GCA_002355295.1_568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>572</td>\n",
       "      <td>NZ_CP023465.1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>6169153</td>\n",
       "      <td>6169291</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR3990735_sra_calls.gff,SRR3990750_sra_calls....</td>\n",
       "      <td>3302,3303,3304,3305,3306,3307</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>GCA_002355295.1_573</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sequence feature    start      end strand  \\\n",
       "2    NZ_CP023465.1   ncRNA    14319    14514      +   \n",
       "17   NZ_CP023465.1   ncRNA   164658   164831      +   \n",
       "26   NZ_CP023465.1   ncRNA   249532   249705      -   \n",
       "37   NZ_CP023465.1   ncRNA   328878   329088      +   \n",
       "38   NZ_CP023465.1   ncRNA   329138   329442      +   \n",
       "..             ...     ...      ...      ...    ...   \n",
       "536  NZ_CP023465.1   ncRNA  5655550  5655801      +   \n",
       "537  NZ_CP023465.1   ncRNA  5655862  5656131      +   \n",
       "550  NZ_CP023465.1   ncRNA  5807459  5807607      -   \n",
       "567  NZ_CP023465.1   ncRNA  6074106  6074294      -   \n",
       "572  NZ_CP023465.1   ncRNA  6169153  6169291      -   \n",
       "\n",
       "                                            file_names  \\\n",
       "2    SRR3990742_sra_calls.gff,SRR3990750_sra_calls....   \n",
       "17   SRR3990749_sra_calls.gff,SRR3990734_sra_calls....   \n",
       "26   SRR3990742_sra_calls.gff,SRR3990749_sra_calls....   \n",
       "37   SRR3990749_sra_calls.gff,SRR3990750_sra_calls....   \n",
       "38   SRR3990749_sra_calls.gff,SRR3990742_sra_calls....   \n",
       "..                                                 ...   \n",
       "536  SRR3990734_sra_calls.gff,SRR3990735_sra_calls....   \n",
       "537  SRR3990735_sra_calls.gff,SRR3990733_sra_calls....   \n",
       "550  SRR3990750_sra_calls.gff,SRR3990734_sra_calls....   \n",
       "567  SRR3990749_sra_calls.gff,SRR3990734_sra_calls.gff   \n",
       "572  SRR3990735_sra_calls.gff,SRR3990750_sra_calls....   \n",
       "\n",
       "                                           row_numbers  prop_overlap  \\\n",
       "2                                 18,19,20,21,22,23,24      0.856410   \n",
       "17                                 128,129,130,131,132      1.000000   \n",
       "26                             170,171,172,173,174,175      1.000000   \n",
       "37                 245,246,247,248,249,250,251,252,253      0.709524   \n",
       "38                                 254,255,256,257,258      0.361842   \n",
       "..                                                 ...           ...   \n",
       "536                      3086,3087,3088,3089,3090,3091      1.000000   \n",
       "537  3092,3093,3094,3095,3096,3097,3098,3099,3100,3...      1.000000   \n",
       "550       3178,3179,3180,3181,3182,3183,3184,3185,3186      1.000000   \n",
       "567                                          3283,3284      1.000000   \n",
       "572                      3302,3303,3304,3305,3306,3307      1.000000   \n",
       "\n",
       "     new_feature  number_of_rnaseq_files score                   id  \n",
       "2           True                       7     3    GCA_002355295.1_3  \n",
       "17          True                       5     2   GCA_002355295.1_18  \n",
       "26          True                       6     2   GCA_002355295.1_27  \n",
       "37          True                       9     2   GCA_002355295.1_38  \n",
       "38          True                       5     2   GCA_002355295.1_39  \n",
       "..           ...                     ...   ...                  ...  \n",
       "536         True                       6     6  GCA_002355295.1_537  \n",
       "537         True                      12     2  GCA_002355295.1_538  \n",
       "550         True                       9     2  GCA_002355295.1_551  \n",
       "567         True                       2     2  GCA_002355295.1_568  \n",
       "572         True                       6     3  GCA_002355295.1_573  \n",
       "\n",
       "[80 rows x 12 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2        14319\n",
      "17      164658\n",
      "26      249532\n",
      "37      328878\n",
      "38      329138\n",
      "        ...   \n",
      "536    5655550\n",
      "537    5655862\n",
      "550    5807459\n",
      "567    6074106\n",
      "572    6169153\n",
      "Name: start, Length: 80, dtype: int64\n",
      "2        14514\n",
      "17      164831\n",
      "26      249705\n",
      "37      329088\n",
      "38      329442\n",
      "        ...   \n",
      "536    5655801\n",
      "537    5656131\n",
      "550    5807607\n",
      "567    6074294\n",
      "572    6169291\n",
      "Name: end, Length: 80, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "start = df.loc[:,'start']\n",
    "end = df.loc[:,'end']\n",
    "\n",
    "print(start)\n",
    "print(end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from Bio import SeqIO\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sequence_cleaner(fasta_file, min_length=50, por_n=100, max_length=500):\n",
    "    # Create our hash table to add the sequences\n",
    "    sequences = {}\n",
    "\n",
    "    # Using the Biopython fasta parse we can read our fasta input\n",
    "    for seq_record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        # Take the current sequence\n",
    "        sequence = str(seq_record.seq).upper()\n",
    "        # Check if the current sequence is according to the user parameters\n",
    "        if (\n",
    "            len(sequence) >= min_length\n",
    "            and (float(sequence.count(\"N\")) / float(len(sequence))) * 100 <= por_n\n",
    "            and len(sequence) <= max_length\n",
    "        ):\n",
    "            # If the sequence passed in the test \"is it clean?\" and it isn't in the\n",
    "            # hash table, the sequence and its id are going to be in the hash\n",
    "            if sequence not in sequences:\n",
    "                sequences[sequence] = seq_record.id\n",
    "            # If it is already in the hash table, we're just gonna concatenate the ID\n",
    "            # of the current sequence to another one that is already in the hash table\n",
    "            else:\n",
    "                sequences[sequence] += \"_\" + seq_record.id\n",
    "\n",
    "    # Write the clean sequences\n",
    "\n",
    "    # Create a file in the same directory where you ran this script\n",
    "    with open(\"clear_\" + fasta_file, \"w+\") as output_file:\n",
    "        # Just read the hash table and write on the file as a fasta format\n",
    "        for sequence in sequences:\n",
    "            output_file.write(\">\" + sequences[sequence] + \"\\n\" + sequence + \"\\n\")\n",
    "\n",
    "    print(\"CLEAN!!!\\nPlease check clear_\" + fasta_file)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file=\"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/positive_control.fna\"\n",
    "min_length=50\n",
    "por_n=100\n",
    "max_length=500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for +=: 'builtin_function_or_method' and 'int'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-e7d7fb2c1187>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0moutFile\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/positive_control_filtered.fna\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"a\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mseq_record\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfasta_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0miter\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0msequence\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq_record\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     if (\n",
      "\u001b[0;31mTypeError\u001b[0m: unsupported operand type(s) for +=: 'builtin_function_or_method' and 'int'"
     ]
    }
   ],
   "source": [
    "outFile = open(\"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/positive_control_filtered.fna\", \"a\")\n",
    "for seq_record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    iter += 1\n",
    "    sequence = str(seq_record.seq).upper()\n",
    "    if (\n",
    "            len(sequence) >= min_length\n",
    "            and (float(sequence.count(\"N\")) / float(len(sequence))) * 100 <= por_n\n",
    "            and len(sequence) <= max_length\n",
    "        ):\n",
    "        seq_name = seq_record.id\n",
    "        sequence = seq_record.seq\n",
    "        outFile.write(f\">{seq_name}\\n{sequence}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-56b233fc9757>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0malignDat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/predicted/predicted_genomic_sequence_matches.txt\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelim_whitespace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0malignDat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0malignDat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0malignDat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"details\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"query_id\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0malignDat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"target_contig\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"coord\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0malignDat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdetails\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"/\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpand\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "alignDat = pd.read_csv(\"/Users/thomasnicholson/phd/RNASeq/srna_seqs/version_1/predicted/predicted_genomic_sequence_matches.txt\", header=None, delim_whitespace=True)\n",
    "\n",
    "alignDat = alignDat.iloc[:,[1,3]]\n",
    "alignDat.columns = [\"details\", \"query_id\"]\n",
    "alignDat[[\"target_contig\", \"coord\"]] = alignDat.details.str.split(\"/\", expand = True)\n",
    "alignDat[[\"target_start\", \"target_end\"]] = alignDat.coord.str.split(\"-\", expand = True)\n",
    "alignDat = alignDat[[\"query_id\", \"target_contig\", \"target_start\", \"target_end\"]]\n",
    "alignDat[\"target_start\"] = alignDat[\"target_start\"].astype(str).astype(int)\n",
    "alignDat['target_start'].dtypes\n",
    "alignDat[\"target_end\"] = alignDat[\"target_end\"].astype(str).astype(int)\n",
    "alignDat['target_end'].dtypes\n",
    "alignDat = alignDat.sort_values(by=['target_start'])\n",
    "display(alignDat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['NC_016824.1', 'NC_004578.1', 'NZ_CP014273.1', 'NC_004741.1',\n",
       "       'NC_009800.1', 'NC_009801.1', 'NC_012967.1', 'NZ_CP025268.1',\n",
       "       'NC_016822.1', 'NC_022912.1', 'NZ_CP014272.1', 'NZ_CP022097.1',\n",
       "       'NZ_CP025515.1', 'NZ_CP016634.1', 'NC_017720.1', 'NC_016823.1',\n",
       "       'NC_008463.1', 'NC_002516.2', 'NZ_CP018026.1', 'NC_019393.1',\n",
       "       'NC_016834.1', 'NC_015554.1', 'NC_009786.1', 'NZ_CP003425.1',\n",
       "       'NC_009791.1', 'NZ_CP025517.1', 'NZ_CP015641.1', 'NC_002947.4',\n",
       "       'NC_009788.1', 'NZ_CP003424.1', 'NZ_LT855377.1', 'NC_003277.2',\n",
       "       'NC_009790.1', 'NC_017718.1', 'NC_016833.1', 'NZ_CP025516.1',\n",
       "       'NC_009787.1', 'NZ_CP015527.1', 'NZ_CP011661.1', 'NZ_LT969519.1',\n",
       "       'NZ_CP011658.1', 'NC_022899.1', 'NZ_LT905143.1', 'NZ_CP011662.1',\n",
       "       'NC_021659.1', 'NC_017541.1', 'NZ_LT594096.1', 'NZ_CP019088.1',\n",
       "       'NZ_CP011961.1', 'NC_006834.1', 'NC_009084.1', 'NZ_CP020358.1',\n",
       "       'NC_017719.1', 'NC_009083.1', 'NC_017540.1', 'NC_018220.1',\n",
       "       'NZ_CP016446.1', 'NZ_CP017473.1', 'NZ_CP023465.1', 'NZ_CP011659.1',\n",
       "       'NZ_CP018920.1', 'NC_009789.1', 'NZ_CP018921.1', 'NZ_CP018919.1',\n",
       "       'NZ_LT906474.1', 'NZ_CP020359.1', 'NZ_CP022660.1', 'NC_003197.2',\n",
       "       'NC_016810.1', 'NZ_CP022658.1', 'NZ_LT855376.1', 'NZ_LT969520.1',\n",
       "       'NC_005966.1', 'NZ_LT594095.1', 'NC_009085.1', 'NZ_CP015526.1',\n",
       "       'NC_014259.1', 'NC_004129.6', 'NC_004633.1', 'NZ_CP018321.1',\n",
       "       'NZ_CP017475.1', 'NZ_CP022659.1', 'NZ_CP018922.1', 'NC_004632.1',\n",
       "       'NZ_LT969521.1', 'NC_019394.1', 'NZ_CP020659.1', 'NZ_CP018322.1'],\n",
       "      dtype=object)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "target_contigs = alignDat.target_contig.unique()\n",
    "display(target_contigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GCA_000283715.1_608', 'GCA_000281215.1_63', 'GCA_002208745.1_77',\n",
       "       ..., 'GCA_002072655.1_311', 'GCA_002072655.1_312',\n",
       "       'GCA_002072655.1_314'], dtype=object)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query_ids = alignDat.query_id.unique()\n",
    "display(query_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_004578.1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>target_contig</th>\n",
       "      <th>target_start</th>\n",
       "      <th>target_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>97980</td>\n",
       "      <td>GCA_000281215.1_63</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>1</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>143061</td>\n",
       "      <td>GCA_002208745.1_77</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>1</td>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>143077</td>\n",
       "      <td>GCA_002208745.1_78</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>17</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8731</td>\n",
       "      <td>GCA_000007565.2_1</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>18</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>94392</td>\n",
       "      <td>GCA_000220485.1_557</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>23240</td>\n",
       "      <td>23737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>236913</td>\n",
       "      <td>GCA_900243355.1_877</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>6393548</td>\n",
       "      <td>6393697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>236959</td>\n",
       "      <td>GCA_900243355.1_878</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>6394355</td>\n",
       "      <td>6394504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>236971</td>\n",
       "      <td>GCA_900243355.1_879</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>6394567</td>\n",
       "      <td>6394704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>237017</td>\n",
       "      <td>GCA_900243355.1_880</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>6395014</td>\n",
       "      <td>6395104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>237050</td>\n",
       "      <td>GCA_900243355.1_881</td>\n",
       "      <td>NC_004578.1</td>\n",
       "      <td>6396412</td>\n",
       "      <td>6396538</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3038 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   query_id target_contig  target_start  target_end\n",
       "97980    GCA_000281215.1_63   NC_004578.1             1         169\n",
       "143061   GCA_002208745.1_77   NC_004578.1             1         196\n",
       "143077   GCA_002208745.1_78   NC_004578.1            17         192\n",
       "8731      GCA_000007565.2_1   NC_004578.1            18         195\n",
       "94392   GCA_000220485.1_557   NC_004578.1         23240       23737\n",
       "...                     ...           ...           ...         ...\n",
       "236913  GCA_900243355.1_877   NC_004578.1       6393548     6393697\n",
       "236959  GCA_900243355.1_878   NC_004578.1       6394355     6394504\n",
       "236971  GCA_900243355.1_879   NC_004578.1       6394567     6394704\n",
       "237017  GCA_900243355.1_880   NC_004578.1       6395014     6395104\n",
       "237050  GCA_900243355.1_881   NC_004578.1       6396412     6396538\n",
       "\n",
       "[3038 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "contig = target_contigs[1]\n",
    "print(contig)\n",
    "subsetDat = alignDat.loc[alignDat['target_contig'] == contig]\n",
    "display(subsetDat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('int64')"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsetDat['target_start'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3038\n"
     ]
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlap_vals(subsetDat, overlaps):\n",
    "    dat_len = len(subsetDat.index)\n",
    "    print(dat_len)\n",
    "    overlapping_ids = []\n",
    "    lengths = []\n",
    "    start_val = 0\n",
    "    end_val = 0\n",
    "    for i in range(0,dat_len):\n",
    "        query_val = subsetDat.iloc[i]['query_id']    \n",
    "        new_start_val = min([subsetDat.iloc[i]['target_start'], subsetDat.iloc[i]['target_end']])\n",
    "        new_end_val = max([subsetDat.iloc[i]['target_start'], subsetDat.iloc[i]['target_end']])\n",
    "    #     print(query_val)\n",
    "    #     print(new_start_val)\n",
    "    #     print(new_end_val)    \n",
    "        if end_val > new_start_val:\n",
    "            overlapping_ids.append(query_val)\n",
    "            len_1 = end_val - start_val\n",
    "            len_2 = new_end_val - new_start_val\n",
    "            shortest_seq = min([len_1, len_2])\n",
    "            overlap_start = max([start_val, new_start_val])\n",
    "            overlap_end = min([end_val, new_end_val])\n",
    "            overlap = (overlap_end - overlap_start)/shortest_seq\n",
    "            overlaps.append(overlap)\n",
    "    #         print(overlap)\n",
    "        else:\n",
    "            end_val = new_end_val\n",
    "            start_val = new_start_val\n",
    "            overlapping_ids = [query_val]\n",
    "    return(overlaps)\n",
    "    #     print(overlapping_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "3038\n",
      "406\n",
      "9197\n",
      "8264\n",
      "7594\n",
      "8571\n",
      "8542\n",
      "10381\n",
      "9922\n",
      "8575\n",
      "3050\n",
      "5544\n",
      "3254\n",
      "181\n",
      "6\n",
      "2759\n",
      "2826\n",
      "2106\n",
      "2164\n",
      "22\n",
      "2296\n",
      "156\n",
      "68\n",
      "1\n",
      "60\n",
      "2211\n",
      "4241\n",
      "152\n",
      "4489\n",
      "181\n",
      "181\n",
      "191\n",
      "103\n",
      "435\n",
      "166\n",
      "40\n",
      "141\n",
      "109\n",
      "47\n",
      "1\n",
      "344\n",
      "5649\n",
      "5094\n",
      "5078\n",
      "313\n",
      "4\n",
      "2471\n",
      "2698\n",
      "2464\n",
      "7\n",
      "6257\n",
      "8\n",
      "6\n",
      "5600\n",
      "2083\n",
      "11\n",
      "177\n",
      "19290\n",
      "2\n",
      "80\n",
      "1\n",
      "8\n",
      "4537\n",
      "8474\n",
      "97\n",
      "173\n",
      "5774\n",
      "5768\n",
      "5728\n",
      "5758\n",
      "2872\n",
      "2969\n",
      "2758\n",
      "2363\n",
      "5822\n",
      "2660\n",
      "3051\n",
      "12\n",
      "2121\n",
      "5079\n",
      "35\n",
      "1\n",
      "13\n",
      "13\n",
      "7\n",
      "1672\n",
      "14\n"
     ]
    }
   ],
   "source": [
    "overlapping_ids = []\n",
    "lengths = []\n",
    "start_val = 0\n",
    "end_val = 0\n",
    "overlaps = []\n",
    "for contig in target_contigs:\n",
    "    subsetDat =  alignDat.loc[alignDat['target_contig'] == contig]\n",
    "    overlaps = get_overlap_vals(subsetDat, overlaps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'overlaps' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-4b6fe916512e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdisplay\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moverlaps\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'overlaps' is not defined"
     ]
    }
   ],
   "source": [
    "display(overlaps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'seaborn'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-a84c0541e888>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mseaborn\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0msns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'seaborn'"
     ]
    }
   ],
   "source": [
    "import seaborn as sns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
